tmk_core/protocol/usb_descriptor.h

```diff
@@ -281,6 +294,15 @@ enum usb_endpoints {
 #        define DIGITIZER_IN_EPNUM SHARED_IN_EPNUM
 #    endif
 #endif
+
+#ifdef LAMPARRAY_ENABLE
+    LAMPARRAY_IN_EPNUM = NEXT_EPNUM,
+#    ifdef USB_ENDPOINTS_ARE_REORDERABLE
+#        define LAMPARRAY_OUT_EPNUM LAMPARRAY_IN_EPNUM
+#    else
+    LAMPARRAY_OUT_EPNUM   = NEXT_EPNUM,
+#    endif
```

Comment:
Looks like all HID reports used by the Dynamic Lighting interface are feature reports — in this case the OUT endpoint is not needed (only output reports could be transferred over the OUT endpoint; feature reports are always sent through the `GetReport` and `SetReport` control transfers, never over the IN or OUT endpoints).  So it should be possible to save some resources by removing `LAMPARRAY_OUT_EPNUM` and all related things (if the MCU does not have bidirectional endpoints, this may be especially important to avoid hitting the endpoint limit).

Having output reports going over the OUT endpoint would actually make things easier to implement, because the data that comes over the regular OUT endpoints can be queued and then processed in the QMK main thread, avoiding any issues caused by doing things in interrupt handlers.  However, the spec for the Lighting and Illumination Page actually requires using `GetReport` and `SetReport` (even though it does not actually mention that the reports should be feature reports).

Unfortunately, `LAMPARRAY_IN_EPNUM` cannot be removed, even though the HID report descriptor does not declare any input reports — apparently the HID spec is worded in such a way that having the IN endpoint is mandatory.  However, the resource allocation for that endpoint should be made minimal, because no data transfer will actually happen over it.

Reply:
Removing the out endpoint in a434d20e2464512d5fc1ec20143f0ef9ef20293d broke windows, device wont type, rgb locks up. Reverted for now, will revisit later.